server/src/test/kotlin/com/handcontrol/server/grpc/GrpcTest.kt

```diff
@@ -0,0 +1,301 @@
+package com.handcontrol.server.grpc
+
+import com.handcontrol.server.protobuf.*
+import com.handcontrol.server.repository.CredentialsRepository
+import io.github.majusko.grpc.jwt.data.GrpcHeader
+import io.github.majusko.grpc.jwt.interceptor.AuthClientInterceptor
+import io.grpc.Channel
+import io.grpc.ClientInterceptors
+import io.grpc.ManagedChannelBuilder
+import io.grpc.Metadata
+import io.grpc.stub.MetadataUtils
+import org.assertj.core.api.Assertions.assertThat
+import org.junit.Assert
+import org.junit.jupiter.api.AfterAll
+import org.junit.jupiter.api.Assertions.assertThrows
+import org.junit.jupiter.api.Assertions.assertTrue
+import org.junit.jupiter.api.BeforeAll
+import org.junit.jupiter.api.Test
+import org.junit.jupiter.api.TestInstance
+import org.junit.jupiter.api.extension.ExtendWith
+import org.springframework.beans.factory.annotation.Autowired
+import org.springframework.boot.autoconfigure.EnableAutoConfiguration
+import org.springframework.boot.test.autoconfigure.jdbc.AutoConfigureTestDatabase
+import org.springframework.boot.test.context.SpringBootTest
+import org.springframework.test.context.junit.jupiter.SpringExtension
+import java.time.LocalDateTime
+import java.time.ZoneOffset
+
+@TestInstance(TestInstance.Lifecycle.PER_CLASS)
+@SpringBootTest(webEnvironment = SpringBootTest.WebEnvironment.DEFINED_PORT)
+@ExtendWith(SpringExtension::class)
+@AutoConfigureTestDatabase(replace = AutoConfigureTestDatabase.Replace.NONE)
+@EnableAutoConfiguration
+class GrpcTest(@Autowired
+               val authClientInterceptor: AuthClientInterceptor
+) {
+    private lateinit var channel: Channel
+
+    private val target = "localhost:6565"
+
+    @Autowired
+    lateinit var db: CredentialsRepository
+    //todo use in memory db
+
+
+    @BeforeAll
+    fun register() {
+        val c = ManagedChannelBuilder.forTarget(target).usePlaintext().build();
+        channel = ClientInterceptors.intercept(c, authClientInterceptor)
+        val stub = HandleRequestGrpc.newBlockingStub(channel)
+
+        val registerResponse = stub.registry(Request.LoginRequest.newBuilder().setLogin("login").setPassword("pass").build())
+        assertThat(registerResponse.token != null)
+    }
+
+    @AfterAll
+    fun deleteRegister(){
+        db.deleteById(db.findByLogin("login")!!.id)
+    }
+
+    @Test
+    fun login() {
+        val c = ManagedChannelBuilder.forTarget(target).usePlaintext().build();
+        channel = ClientInterceptors.intercept(c, authClientInterceptor)
+        val stub = HandleRequestGrpc.newBlockingStub(channel)
+
+        val loginResponse = stub.login(Request.LoginRequest.newBuilder().setLogin("login").setPassword("pass").build())
+        assertThat(loginResponse.token != null)
```

Comment:
all of this assertions are redundant because, for example, response can't be null. You should check that the response is not error and the answer is correct in successful case. And is error on fail, you have no tests for that case.

Reply:
Most importance part of the login is token, so i check if it return token with a exist account.
Test for not exist account is below.